feat: DatabaseRules (de)serialisation support

domodwyer opened the pull request:
Allows Conductor to serialise these structures into JSON.

---

* feat: DatabaseRules (de)serialisation support (7d3b6bf8)

      Derives serde::Serialise & Deserialise for the DatabaseRules struct and 
      children.

mkmik commented on data_types/src/database_rules.rs:
```diff
@@ -724,4 +727,11 @@ mod tests {
+
+    #[test]
+    fn test_assert_serde() {
+        fn ensure<'de, T: Serialize + Deserialize<'de>>(_t: T) {}
```
👍

tustvold commented:
Will conductor be requiring this serialization to be stable? If so it might be better to use the protobuf serialization which is what is actually persisted?

domodwyer commented:
Amusingly I just fixed the breaking gRPC changes when I updated the IOx pin 😂  Is the gRPC interface more stable? If we add a new config parameter, surely it'll have to go in both?

Additionally using JSON allows us to store it in postgres as semi-structured, queryable blobs with the JSONB data type rather than an opaque bytes blob if we use proto - we could potentially JSON encode the gRPC type if the extra hops are worth it.

tustvold commented:
> Is the gRPC interface more stable?

It is supposed to be, but I think we're running a bit fast and loose at the moment...

>  surely it'll have to go in both

Not necessarily, the DatabaseRules is intended to be the IOx internal representation of the configuration, complete with any necessary defaulting, and not including any deprecated/removed fields. Once we start maintaining API compatibility properly, I'd imagine the schema conversion logic would exist prior to DatabaseRules conversion. 

More generally supporting schema migration using serde makes me somewhat apprehensive, there aren't good tools that I'm aware of to verify it, renaming/adding fields is a breaking change, etc...

I don't have particularly strong opinions in this space, but my past experience of not having a clear separation between API types and internal types has resulted in a mess, with different bits of the system applying different defaulting / schema migration logic.

domodwyer commented:
There are already issues with defaulting - I was going to patch it up in a separate issue/PR as it has wider implications - currently we set `max_active_compactions` to the number of CPUs which is great, unless the `Default` impl is called on some machine other than the server. There's also no clear way to send a request without specifying `max_active_compactions` as it is not optional, so we can never ask the server to set it to the appropriate value via the generated rust gRPC types. This is easily fixable with an `optional` proto field 👍 

I also agree with not-leaking-your-internal-types, but my understanding from @pauldix was these items in `data_types` _were intended to be shared_ to reduce boilerplate between the two projects (and other crates). At least that was the initial idea, are we saying that is no longer the case?

tustvold commented:
> unless the Default impl is called on some machine other than the server

This is precisely the kind of thing that makes me apprehensive about using these types as part of a public API, this (and a number of other things) are intended to be server-side defaults. Client side defaulting is often problematic, but using data_types as currently written forces defaulting on decode. This gets even more mad if you convert on the client side to data_types from the protobuf returned by a server - it will default the response!

>  these items in data_types were intended to be shared to reduce boilerplate between the two projects (and other crates)

This was the original intent, but I was under the impression we had reached a concensus when the gRPC API was introduced that data_types would not be wire-compatible, but perhaps we hadn't. 

All I care about is:

* The types used internally by IOx aren't burdened by the need to allow API-evolution - if current IOx needs a field it doesn't need to be `Option` and can just be defaulted in one place, if a field doesn't need to exist anymore it can just be removed, etc...
* We don't lose the flexibility afforded by protobuf's wire-format to evolve the API without breaking older/newer clients

How that is achieved, I don't particularly mind. I had envisaged that clients would just consume the generated protobuf types, as these represent exactly what the server/client sent on the wire, but if there are other ways to achieve the above - that sounds fine to me :grin:

domodwyer commented:
> it will default the response

If the server is applying default values, they will be returned populated in the response so no worries.

> if current IOx needs a field it doesn't need to be Option and can just be defaulted in one place

I think you've misunderstood - I'm taking about the generated `CreateDatabase` gRPC request type. it's not an optional field, so there's no way to say "I would like the server to apply the server side default" - the client is forced to specify a value. I assume this isn't strictly true, and there's actually some mapping from 0 -> server side default, but it would be much nicer if the API expressed these optionally configurable semantics through the type system instead of per-field magic values. This is obviously a different issue though 👍 

> We don't lose the flexibility afforded by protobuf's wire-format to evolve the API without breaking older/newer clients

Using JSON would effectively the same as storing protobuf encodings - neither are sent directly over the wire, but instead decoded to an IOx type within Conductor. However, the JSON format is queryable, which makes for a more flexible solution. Any discrepancy between conductor & IOx's API wire types (which remain proto) are resolved by the proto backwards compat anyway. I don't think this is an issue either.

tustvold commented:
>  I don't think this is an issue either.

Provided IOx and Conductor are always updated in lock-step, I think I agree. There are probably ways you can end up in a chicken-and-egg upgrade, but I guess we cross that bridge when it comes

domodwyer commented:
> Provided IOx and Conductor are always updated in lock-step

This certainly isn't true, it would make managing IOx updates impossible!

If you changed a type in the `DatabaseRules`, and I maintained my own de-coupled version of `DatabaseRules` my type would still serialise to the same proto it did before your change until I update my type. If your `DatabaseRules` change is wire-compatible with the old proto (which it should be) there is no problem, old and new communicate fine.

Next time I update the IOx pin to be able to use your new config I'd have to update my type too, which in turn, is no longer JSON compatible with the old one and some sort of migration would have to take place. I don't see how this is any different from using `DatabaseRules` directly - de-coupling in this instance seems both redundant, and given the complexity of the types, error-prone. If Conductor were less tightly coupled to IOx, it probably would make sense.

I'm happy to explain in a zoom call or something if you want 👍

tustvold commented:
>  some sort of migration would have to take place.

Yeah, I guess I was just thinking if you used the protobuf this wouldn't be necessary, if you'd rather store JSON and have to write migrations for it - that's your call :grinning:

domodwyer commented:
"Migrations" in this context are a `serde::deserialize_with` annotation and fn - unless your daily workflow involves changing a type in `DatabaseRules` every morning it isn't going to be a burden worth optimising for ;)
